src/meta/src/hummock/compaction_scheduler.rs

```diff
@@ -368,6 +318,207 @@ where
+        &self,
+        sched_channel: Arc<CompactionRequestChannel>,
+        shutdown_rx: Shared<Receiver<()>>,
+        mut compaction_selectors: HashMap<compact_task::TaskType, Box<dyn LevelSelector>>,
+        event_stream: impl Stream<Item = SchedulerEvent>,
+    ) {
+        use futures::pin_mut;
+        pin_mut!(event_stream);
+
+        loop {
+            let item = futures::future::select(event_stream.next(), shutdown_rx.clone()).await;
+            match item {
+                Either::Left((event, _)) => {
+                    if let Some(event) = event {
+                        match event {
+                            SchedulerEvent::Channel((compaction_group, task_type)) => {
+                                // recv
+                                if !self
+                                    .on_handle_compact(
+                                        compaction_group,
+                                        &mut compaction_selectors,
+                                        task_type,
+                                        sched_channel.clone(),
+                                        shutdown_rx.clone(),
+                                    )
+                                    .await
+                                {
+                                    break;
+                                }
+                            }
+                            SchedulerEvent::DynamicTrigger => {
+                                // Disable periodic trigger for compaction_deterministic_test.
+                                if self.env.opts.compaction_deterministic_test {
+                                    continue;
+                                }
+                                // Periodically trigger compaction for all compaction groups.
+                                self.on_handle_trigger_multi_grouop(
+                                    sched_channel.clone(),
+                                    compact_task::TaskType::Dynamic,
+                                )
+                                .await;
+                                continue;
+                            }
+                            SchedulerEvent::SpaceReclaimTrigger => {
+                                // Disable periodic trigger for compaction_deterministic_test.
+                                if self.env.opts.compaction_deterministic_test {
+                                    continue;
+                                }
+                                // Periodically trigger compaction for all compaction groups.
+                                self.on_handle_trigger_multi_grouop(
+                                    sched_channel.clone(),
+                                    compact_task::TaskType::SpaceReclaim,
+                                )
+                                .await;
+                                continue;
+                            }
+                            SchedulerEvent::TtlReclaimTrigger => {
```

Comment:
Will TtlReclaimTrigger and SpaceReclaimTrigger conflicts with each other? Under default config, space reclaim (interval 60min) happens on every two TTL reclaim (30min). Since both space and ttl reclaim check L6 SSTs, is it possible for one to win over the other and cause starvation? Space reclaim and ttl recliam have similar picker logic. This keeps me thinking that whether we should only keep one general picker and trigger for reclaim considering both table drop and ttl.

Reply:
No. Each of them would exit if they found some sst files has been selected.

Reply:
With the existing implementation, only the contiguity of the files in the selection task is required, not the whole level, i.e. SpaceReclaim alternates with TtlReclaim and does not cause starvation due to pending status.